### ISSUE TITLE
main content link styling for accessibility

### ISSUE THREAD

**Issue by Alex-Jordan** [rbeezer/mathbook#1003]
Everything here is about "true links", not knowl links 
(whose accessibility merits a full investigation later). 
At least as of 2/22/19, morning, I think we are not styling links accessibly. 

For reference, at: 
https://pretextbook.org/examples/sample-article/html/text-in-paragraphs.html#section-urls,

you can see an xref link (search for "Subsection"), 
an external link that just prints the href, 
and an external link where the author supplied link text. 

Here is a WCAG article on styling links:
https://webaim.org/techniques/hypertext/link_text#appearance

Note there are different "rules" for links in main content 
than for links in a nav bar, ToC, etc. 
I am only posting about main content links here.
For a link in main content, the "safest" most accessible thing
is to leave it underlined. We are not doing that, which can be OK, 
if some additional rules are followed.

1. The link text must have a 3:1 contrast ratio from the surrounding non-link text.

2. The link must present a "non-color designator" (typically the introduction of the underline) on both mouse hover and keyboard focus.

We are not doing well with item 1. Here is a contrast checker: 
https://webaim.org/resources/contrastchecker/

For xref links (muddy reddish), our contrast ratio is currently 2.35:1. 
For an external link with author-provided text (a deep blue), 
the ratio is only 1.88:1.  Same for "href" links, 
but it is not clear if the background "code" coloring mitigates that.

We are doing OK with item 2, but could be made better.
It looks like in all three types of link, focus brings a dotted outline. 
And hover does some kind of highlighting. 
Except in the case of an external link that just prints the href, 
the highlighting is not very discernible
because of the background "code" color. 
It might be good if hover could also change something 
about the background "code" color for such links.

There is a note at the end of the article that suggests
the focus and hover effects should be the same. 
Maybe the outlining I see for focus is just the browser default, 
and simply adding `a:focus` to that part of the CSS 
where the hover effect is defined would be a simple improvement.

**Comment by Alex-Jordan** [comment]
By the way,

1. On the one hand, David is keeper of the CSS.
2. On the other hand, I don't mean this to be a "David, please fix this" post. I haven't proposed a full fix, specifically with identifying color replacements. That's a significant bit of research and experimenting to do.
3. On my third hand, while I can choose colors and determine colors with good ratios, I am not a good person to choose colors that are pleasing. Just ask my wife with her comments after I dress myself sometimes. I have the red-green colorblindness gene, and it affects what I perceive as "pleasant" and what I perceive (or do not perceive) to "clash".

At a minimum, I just want to document the issue, since I know I can do that part.

**Comment by davidfarmer** [comment]
I am happy to work on this, but need to make sure I am doing the right 
thing.

The muddy-red knowl link has color  #932919

At the contrast checker link the contrast ratio of this with #000000,
the black main text, is 2.57:1.  I realize that is not high enough,
but you wrote 2.35:1.  How did you get the 2.35?

That happens to be the contrast ratio with #0f0f0f, but maybe that is
not meaningful.

**Comment by Alex-Jordan** [comment]
For me, the color is #900. My browser indicates it gets this from pretext_add_on.css:

```
.pretext-content a.internal {
    color: #900;
}
```

Is it a matter of having visited the link and that makes your version slighltly lighter?

**Comment by davidfarmer** [comment]
Note that it is almost impossible to have a color that is 7:1 contrast
with white and 3:1 contrast with black.  So, without a decoration on
the links we cannot achieve AAA accessibility.

The color for internal links is now  #A62E1C , which has ratios
with black and white, 3.03 and 6.93, respectively.

And external links are now  #2B5F82 , with ratios 3.06 and 6.86 .

I also changed the background colors when hovering or focus,
and added the underline in both cases.

I also lightened the background for "code" when it is in a link,
as in:

------
And we include a no-content version of the same link:  <a class="external" 
href="https://en.wikipedia.org/wiki/Carleson%27s_theorem" 
target="_blank"><code class="code-inline 
tex2jax_ignore">https://en.wikipedia.org/wiki/Carleson%27s_theorem</code></a>
------

(I ask why that is "code"?  Maybe it is code, but what is an appropriate
way to style code that is a link?  Should we make a distinction between
code in the more naive sense it is usually used, and a URL?)

There is a thin outline that the browser puts around an "a" with focus.
That is not set by CSS.  Other than that, I think I styled the links
the same for hover and focus.

Unless I missed something, this can be closed.

**Comment by rbeezer** [comment]
Thanks, Alex, for the research, and David for the quick action.

I added a bit to the Author's Guide explaining this while doing #1005.  Then got carried away talking more generally about links.  **Added** three new footnotes.  Alex might want to review changes for correctness.  At 4d8e0fa8884e673ae004c33408b55fea1c031840

I think this one has served its purpose, so maybe Alex wants to close it now, even if there may be some fine-tuning.  Similar work for knowls could be a new issue.
